### PR TITLE
[WIP] Implement yul to coreyul translation

### PR DESCRIPTION
**Pull request by sayon**
This is an ongoing work in progress/prototype.

# What ❔

Implement a transpiler from YUL to EasyCrypt in a separate binary.
EasyCrypt spec: https://cs-people.bu.edu/gaboardi/teaching/22-chalmers-course/easycrypt-refman.pdf

## Why ❔

A part of formal verification effort.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,6 +9,10 @@ edition = "2021"
 description = "EraVM Solidity compiler"
 links = "compiler-rt-zksolc"
 
+[[bin]]
```
Comment (hedgar2017) [inline Cargo.toml:12]: Please move below `zksolc` so it can be seen as the main binary.

File: src/easycrypt/syntax/definition.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    /// Location of the original variable in the source YUL file.
+    pub location: Option<crate::yul::path::Path>,
+
+    /// Type of the variable, if the definition is annotated with one.
+    pub r#type: Option<Type>,
+}
+
+impl Definition {
+    /// Produce a reference to the variable.
```
Comment (hedgar2017) [inline src/easycrypt/syntax/definition.rs:23]: Please use `///` surrounding lines for structures and functions.

File: src/easycrypt/syntax/proc/name.rs [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+//! Name of a procedure in EasyCrypt, which can be a user-defined custom name or
+//! one of the pre-defined names such as `mstore`.
+//!
+
+use std::fmt::Display;
+
+use crate::easycrypt::syntax::Name;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ProcName {
```
Comment (hedgar2017) [inline src/easycrypt/syntax/proc/name.rs:11]: Could the enum from the Yul parser be reused?

File: src/easycrypt/syntax/statement/if_conditional.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+use crate::easycrypt::syntax::statement::Statement;
+
+/// EasyCrypt AST node containing an `if` statement.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IfConditional {
+    /// On which condition the statement should be executed.
+    pub condition: Expression,
+    /// Statement to be executed if the condition is true.
+    pub yes: Box<Statement>,
+    /// Optionally, a statement to be executed if the condition is false.
```
Comment (hedgar2017) [inline src/easycrypt/syntax/statement/if_conditional.rs:15]: IIRC there is no `else` in Yul. Does EasyCrypt introduce it?

File: src/easycrypt/syntax/statement/mod.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+use self::block::Block;
+use self::call::ProcCall;
+use self::if_conditional::IfConditional;
+use self::while_loop::WhileLoop;
+
+use crate::easycrypt::syntax::definition::Definition;
+use crate::easycrypt::syntax::expression::Expression;
+use crate::easycrypt::syntax::reference::Reference;
+
+/// EasyCrypt AST nodes containing statements. Statements are a syntactic
```
Comment (hedgar2017) [inline src/easycrypt/syntax/statement/mod.rs:21]: Well, if EasyCrypt is not a strict subset of Yul, it makes sense to redefine the whole AST.
Is it correct? Could you add the link to the AST spec to PR description?

File: src/yul-easycrypt/arguments.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    /// Create one file per component and contract/file at the specified directory, if given.
+    #[structopt(short = "o", long = "output-dir")]
+    pub output_directory: Option<PathBuf>,
+
+    /// Enable system contract compilation mode.
+    /// In this mode EraVM extensions are enabled. For example, calls to addresses `0xFFFF` and below
+    /// are substituted by special EraVM instructions.
+    /// In the Yul mode, the `verbatim_*` instruction family is available.
+    #[structopt(long = "system-mode")]
```
Comment (hedgar2017) [inline src/yul-easycrypt/arguments.rs:30]: Let's turn it into `--enable-eravm-extensions`, like in [this PR](https://github.com/matter-labs/era-compiler-solidity/blob/e6007464b907ee0916b6e85fd062410bb0583a17/src/zksolc/arguments.rs#L157).

File: src/yul-easycrypt/main.rs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        Ok(()) => era_compiler_common::EXIT_CODE_SUCCESS,
+        Err(error) => {
+            eprintln!("{error}");
+            era_compiler_common::EXIT_CODE_FAILURE
+        }
+    })
+}
+
+fn print_version() {
+    println!(
```
Comment (hedgar2017) [inline src/yul-easycrypt/main.rs:25]: Can `structopt` print this out of the box?
`zksolc` does that IIRC.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
We also need unit tests!

File: src/easycrypt/syntax/statement/mod.rs (hunk 5 above)
Comment (sayon) [reply src/easycrypt/syntax/statement/mod.rs:22]: It is not a subset; for once, there are functions and procedures, there are binary operations, `else`, no `for` loop but `while` loop etc.
I have added a link to the EasyCrypt reference to the PR and once we move the code into a separate repository I ll add a short description of the supported AST there directly.

File: src/yul-easycrypt/main.rs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    let arguments = Arguments::new();
+    arguments.validate()?;
+
+    if arguments.version {
+        print_version();
+        return Ok(());
+    }
+
+    let input_files = arguments.input_files_paths()?;
+    let _suppressed_warnings = match arguments.suppress_warnings {
```
Comment (hedgar2017) [inline src/yul-easycrypt/main.rs:36]: Please remove unused CLI parameters!

File: src/yul-easycrypt/main.rs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+///
+/// The auxiliary `main` function to facilitate the `?` error conversion operator.
+///
+fn main_inner() -> anyhow::Result<()> {
+    let arguments = Arguments::new();
+    arguments.validate()?;
+
+    if arguments.version {
+        print_version();
```
Comment (hedgar2017) [inline src/yul-easycrypt/main.rs:31]: Can it be done without this function like in other executables?

File: src/project/contract/mod.rs [hunk 10]
```diff
@@ -373,4 +375,14 @@ impl Contract {
     pub fn get_missing_libraries(&self) -> HashSet<String> {
         self.ir.get_missing_libraries()
     }
+
+    ///
+    /// Get the underlying Yul object, if available
+    ///
+    pub fn get_yul_object(&self) -> Option<&Object> {
+        match &self.ir {
+            IR::Yul(y) => Some(&y.object),
```
Comment (hedgar2017) [inline src/project/contract/mod.rs:384]: Can the whole thing be eliminated?

File: src/project/contract/mod.rs [hunk 11]
```diff
@@ -21,6 +21,8 @@ use crate::solc::version::Version as SolcVersion;
 use self::ir::IR;
 use self::metadata::Metadata;
```
Comment (hedgar2017) [inline src/project/contract/mod.rs:22]: Is metadata unused?

File: src/easycrypt/translator/expression/function_call.rs [hunk 12]
```diff
@@ -0,0 +1,190 @@
+//!
+//! Transpilation of YUL function calls.
```
Comment (hedgar2017) [inline src/easycrypt/translator/expression/function_call.rs:2]: Please replace `YUL` => `Yul` for unification.

File: Cargo.toml [hunk 13]
```diff
@@ -40,6 +44,7 @@ zkevm-assembly = { git = "https://github.com/matter-labs/era-zkEVM-assembly", br
 
 era-compiler-common = { git = "https://github.com/matter-labs/era-compiler-common", branch = "main" }
 era-compiler-llvm-context = { git = "https://github.com/matter-labs/era-compiler-llvm-context", branch = "main" }
+topological-sort = "0.2.2"
```
Comment (hedgar2017) [inline Cargo.toml:47]: Please group with other 3rd-party deps.

File: Cargo.toml [hunk 14]
```diff
@@ -40,6 +44,7 @@ zkevm-assembly = { git = "https://github.com/matter-labs/era-zkEVM-assembly", br
...
 
 [dependencies.inkwell]
 git = "https://github.com/matter-labs-forks/inkwell"
```
Comment (hedgar2017) [inline Cargo.toml:50]: Did you plan to remove LLVM deps from `easycrypt`?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

**Comment by sayon** [comment]
development moved to: https://github.com/matter-labs/yul-easycrypt

**Pull request closed without merging**